### PR TITLE
Reopen completed tickets on explicit follow-up sends

### PR DESCRIPTION
**Pull request by morapelker**
## Summary
- Reopen `done` and `merged` tickets to `in_progress` after explicit follow-up messages.
- Distinguish user sends from status replays, streaming updates, and background task notifications.
- Add session-sync coverage for reopening and non-reopening scenarios.

## Testing
- Added unit tests covering explicit sends, stale sends, terminal prompts, task notifications, and review tickets.

---

> [!NOTE]
> **Medium Risk**
> Touches core kanban column rules and session-status→board coordination with many edge cases (stamps, replays, multi-project load order); broad unit test coverage reduces regression risk but behavior is easy to get wrong in production.
> 
> **Overview**
> **Done** and **merged** kanban tickets can move back to **in_progress** when the user genuinely resumes work (follow-up message, plan implement, queued drain, CLI prompt)—not when the session merely replays `working` or auto-resumes from task notifications.
> 
> `useWorktreeStatusStore` now classifies `working`/`planning` transitions: unconsumed `userExplicitSendTimes` (one-shot per send), `markNextWorkingStatusExplicit` for queue drains, `UserPromptSubmit` hooks, and CLI plan follow-ups count as explicit; `taskNotification` and stale stamps do not. Kanban sync only reopens terminal columns on `session_working` with `explicitSend: true`, or on `implement` (with `auto_approve_plan` / archived guards).
> 
> **Unload recovery:** pending explicit/implement reopen markers are stored per session and applied in `reconcileFinishedSessions` when a project’s tickets load late; `status_cleared` and finished runs drop stale markers. **Implement rollback:** failed plan approval/dispatch rolls terminal tickets back via `moveTicket(..., { skipCompletionEffects: true })` unless a newer send superseded the attempt. `sendFollowupToSession` gains `skipSendBookkeeping` so implement paths don’t double-stamp status.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0abbe87cd9e168d1d7b8f3857090c028d2453f9e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/renderer/src/stores/useWorktreeStatusStore.ts [hunk 1]
```diff
@@ -148,7 +150,18 @@ export const useWorktreeStatusStore = create<WorktreeStatusState>((set, get) =>
...
+      // immediately follows a user-initiated send (every send path writes
+      // userExplicitSendTimes right before flipping the status) or carries a
+      // UserPromptSubmit hook event that isn't a background task-notification
+      // auto-resume (prompts typed straight into a CLI terminal). Streaming
+      // re-emits and status replays match neither, so they can never pull a
+      // ticket out of done/merged.
+      const explicitSendAt = userExplicitSendTimes.get(sessionId)
+      const explicitSend =
+        (explicitSendAt !== undefined && Date.now() - explicitSendAt < 15_000) ||
+        (metadata?.hookEventName === 'UserPromptSubmit' && metadata.taskNotification !== true)
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts]: Fixed in d81c731d — the send stamp is now consumed by the first working/planning transition it explains (tracked in a separate map so the elapsed timer keeps `userExplicitSendTimes` intact), and `taskNotification: true` events are excluded before the timestamp branch, so they can neither reopen a ticket nor consume a pending stamp.

File: src/renderer/src/stores/useWorktreeStatusStore.ts [hunk 2]
```diff
@@ -148,7 +150,18 @@ export const useWorktreeStatusStore = create<WorktreeStatusState>((set, get) =>
...
+      notifyKanbanSessionSync(sessionId, { type: 'session_working', explicitSend })
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts]: Fixed in d81c731d — each send stamp now marks at most one working transition as explicit (one-shot consumption via a separate consumed-marker map), so re-emits inside the 15s window can no longer reopen a ticket the user moved back to done/merged. Covered by a regression test.

[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts]: Fixed in d81c731d — rather than stamping the approve paths, the kanban `implement` event (which only fires on genuine user plan approvals, both opencode and CLI) now reopens done/merged tickets to in_progress directly. Review/todo tickets keep moving via session_working as before. Covered by regression tests.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 3]
```diff
@@ -1123,6 +1126,16 @@ export const useKanbanStore = create<KanbanState>()(
                     })
                     .catch(() => {})
                 }
+                // Approving a plan on a done/merged ticket resumes real work,
+                // but its working status carries no send stamp (opencode) or a
+                // PostToolUse hook (CLI), so session_working won't reopen it.
+                // implement only fires on genuine user approvals — reopen here.
+                // Other columns keep moving via session_working as before.
+                if (ticket.column === 'done' || ticket.column === 'merged') {
+                  get()
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts:1217]: Fixed in 8becec06 — both implement paths that wrote the send stamp after `setSessionStatus('working')` (KanbanTicketModal.handleImplement and the claude-code branch of SessionView.handlePlanReadyImplement) now write it before, so the implement's own working transition consumes the stamp immediately and no unconsumed marker survives for later replays. Audited every other `userExplicitSendTimes.set` site — all already stamp before the status flip.

File: src/renderer/src/stores/useWorktreeStatusStore.ts [hunk 4]
```diff
@@ -148,7 +157,27 @@ export const useWorktreeStatusStore = create<WorktreeStatusState>((set, get) =>
...
+      // explicit, and each send stamp is consumed by the first working
+      // transition it explains, so streaming re-emits and status replays can
+      // never pull a ticket out of done/merged.
+      let explicitSend = false
+      if (metadata?.taskNotification !== true) {
+        const explicitSendAt = userExplicitSendTimes.get(sessionId)
+        const hasUnconsumedSend =
+          explicitSendAt !== undefined &&
+          Date.now() - explicitSendAt < 15_000 &&
+          consumedExplicitSendTimes.get(sessionId) !== explicitSendAt
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts:192]: Fixed in c3afcf0a — queued follow-ups now count as explicit via a one-shot `markNextWorkingStatusExplicit` marker consumed by the drain's working transition, called in both the active-session (SessionView) and background (useOpenCodeGlobalListener) idle-drain paths. Kept `userExplicitSendTimes` untouched so the elapsed timer doesn't restart at dispatch. Task-notifications can't consume the marker; covered by store tests.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 5]
```diff
@@ -2227,10 +2227,13 @@ function PlanReviewModeContent({
...
       await useSessionStore.getState().setSessionMode(sessionId, 'build')
       lastSendMode.set(sessionId, 'build')
-      useWorktreeStatusStore.getState().setSessionStatus(sessionId, 'working')
+      // Send bookkeeping must precede setSessionStatus so the working
+      // transition it causes consumes the explicit-send stamp — a stamp left
+      // unconsumed would mark a later status replay as an explicit send.
       messageSendTimes.set(sessionId, Date.now())
       userExplicitSendTimes.set(sessionId, Date.now())
       snapshotTokenBaseline(sessionId)
+      useWorktreeStatusStore.getState().setSessionStatus(sessionId, 'working')
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx:2283]: Fixed in c3afcf0a — handleImplement now restores the original done/merged column (and clears the dangling working status on the missing-path return) on all three failure paths: approve error, working-path-not-found, and background implementation-send failure. The optimistic reopen is kept for the success path per the existing review-column UX pattern.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 6]
```diff
@@ -1138,13 +1151,22 @@ export const useKanbanStore = create<KanbanState>()(
...
                     .catch(() => {})
                 }
-                if (ticket.column === 'todo' || ticket.column === 'review') {
+                const reopenedByFollowup =
+                  event.explicitSend === true &&
+                  (ticket.column === 'done' || ticket.column === 'merged')
+                if (
+                  ticket.column === 'todo' ||
+                  ticket.column === 'review' ||
+                  reopenedByFollowup
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts:1254]: Declining this one as by-design. The optimistic move-before-delivery is the established pattern for these paths — the pre-existing review→in_progress behavior works exactly the same way (ReviewModeContent moves the ticket FIRST for immediate UI feedback and only toasts on send failure), and claude-cli PTY delivery is fire-and-forget with no ack to defer on. A failed follow-up shows an error toast and either retries via the requeue or is one drag away from done/merged, while wiring column-capture + rollback through every composer/modal/drain send path would be a large risk for a rare, user-visible failure. The case where this genuinely mattered — implement reopening a ticket before approval succeeded — was fixed in c3afcf0a.

File: src/renderer/src/stores/useWorktreeStatusStore.ts [hunk 7]
```diff
@@ -148,7 +157,27 @@ export const useWorktreeStatusStore = create<WorktreeStatusState>((set, get) =>
...
+        if (hasUnconsumedSend) {
+          consumedExplicitSendTimes.set(sessionId, explicitSendAt)
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts:194]: Fixed in 2f56e342 — syncTicketWithSession now records explicit sends that matched no loaded ticket, and reconcileFinishedSessions (already run after every ticket load) attributes them: a done/merged ticket whose session is still working/planning reopens on load, one-shot, and the pending entry is cleared whenever a loaded ticket handles a send live so it can't resurrect a re-done ticket. Sends whose session already finished by load time stay terminal, matching the existing reconcile philosophy of acting only on positive evidence.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 8]
```diff
@@ -1123,6 +1126,16 @@ export const useKanbanStore = create<KanbanState>()(
...
                     .catch(() => {})
                 }
+                // Approving a plan on a done/merged ticket resumes real work,
+                // but its working status carries no send stamp (opencode) or a
+                // PostToolUse hook (CLI), so session_working won't reopen it.
+                // implement only fires on genuine user approvals — reopen here.
+                // Other columns keep moving via session_working as before.
+                if (ticket.column === 'done' || ticket.column === 'merged') {
+                  get()
+                    .moveTicket(ticket.id, projectId, 'in_progress', ticket.sort_order)
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts]: Fixed in 2f56e342 — the implement reopen is now gated on !ticket.auto_approve_plan, so a hook-server auto-approved ExitPlanMode leaves an armed done/merged ticket in place. Manual approvals still reopen: the modal and plan-card paths stamp userExplicitSendTimes, so their working transition reopens via the explicit-send branch independently of the implement event.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 9]
```diff
@@ -2219,6 +2219,19 @@ function PlanReviewModeContent({
...
+    // implementation fails, put the ticket back in its terminal column — an
+    // in_progress ticket with no running session is a lie on the board.
+    const columnBeforeImplement = ticket.column
+    const restoreTerminalColumn = (): void => {
+      if (columnBeforeImplement !== 'done' && columnBeforeImplement !== 'merged') return
+      useKanbanStore
+        .getState()
+        .moveTicket(ticket.id, ticket.project_id, columnBeforeImplement, ticket.sort_order)
+        .catch(() => {})
+    }
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx:2266]: Fixed in ae4eca59 — restoreTerminalColumn now reads the ticket's current column from the store and only rolls back when it is still in_progress (the state our optimistic reopen created). Any newer transition — user drag or session event — is left untouched.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 10]
```diff
@@ -434,25 +441,39 @@ export const useKanbanStore = create<KanbanState>()(
...
+              get().updateTicket(ticket.id, projectId, { plan_ready: true }).catch(() => {})
+            }
+            get().moveTicket(ticket.id, projectId, 'review', ticket.sort_order).catch(() => {})
+          } else if (ticket.column === 'done' || ticket.column === 'merged') {
+            // Recover an explicit follow-up reopen that fired before these
+            // tickets were loaded. Only while the session is still actively
+            // running — a session that already finished (or never started)
+            // gives no license to leave the terminal column.
+            if (!pendingExplicitReopens.has(ticket.current_session_id)) continue
+            if (status !== 'working' && status !== 'planning') continue
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts]: Fixed in 4a32d08f — pending entries are now dropped as soon as the originating run stops being active (session_completed / plan_ready / session_error all clear them, independent of whether any ticket was loaded), so a stale marker can't pair with a later unrelated working status like a background auto-resume. Covered by a regression test.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 11]
```diff
@@ -1154,6 +1206,12 @@ export const useKanbanStore = create<KanbanState>()(
             }
           }
         }
+        // Remember explicit sends that found no loaded ticket so the reopen
+        // can be recovered when the project's tickets load (reconcile below).
+        if (event.type === 'session_working' && event.explicitSend === true) {
+          if (matchedTicket) pendingExplicitReopens.delete(sessionId)
+          else pendingExplicitReopens.add(sessionId)
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts]: Fixed in 4a32d08f — the registry is now sessionId → set of handled project IDs. An explicit send marks the projects loaded at send time as handled (they reopened live); reconcileFinishedSessions attributes the send per project on later loads, recovering every linked terminal ticket in that project in one pass, and marks the project handled on any linked-ticket visit so a re-done ticket can't be re-reopened by the same send. Covered by a multi-project regression test.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 12]
```diff
@@ -2219,6 +2219,27 @@ function PlanReviewModeContent({
...
+    const restoreTerminalColumn = (): void => {
+      if (columnBeforeImplement !== 'done' && columnBeforeImplement !== 'merged') return
+      // Only undo our own optimistic reopen. The failure can land seconds
+      // later — if the user or a session event moved the ticket again in the
+      // meantime, that newer transition wins.
+      const current = useKanbanStore
+        .getState()
+        .tickets.get(ticket.project_id)
+        ?.find((t) => t.id === ticket.id)
+      if (current?.column !== 'in_progress') return
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx:2259]: Fixed in fcb810d9 — handleImplement records its own messageSendTimes stamp as a generation token (sendFollowupToSession now takes skipSendBookkeeping so its internal re-stamp doesn't clobber the token), and every failure path checks it: when a newer send has bumped the stamp, the late failure neither clears the session status nor restores the terminal column.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 13]
```diff
@@ -1138,13 +1181,22 @@ export const useKanbanStore = create<KanbanState>()(
...
+                // (never for streaming re-emits or status replays).
                 if (ticket.plan_ready) {
                   get()
                     .updateTicket(ticket.id, projectId, { plan_ready: false })
                     .catch(() => {})
                 }
-                if (ticket.column === 'todo' || ticket.column === 'review') {
+                const reopenedByFollowup =
+                  event.explicitSend === true &&
+                  (ticket.column === 'done' || ticket.column === 'merged')
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts]: Fixed in fcb810d9 — tickets with archived_at set are excluded from all three reopen paths (live explicit-send session_working, implement, and reconcile recovery), so an archived done/merged ticket stays in the archive UI with its Unarchive action intact. Covered by regression tests.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 14]
```diff
@@ -434,25 +445,50 @@ export const useKanbanStore = create<KanbanState>()(
...
+              get().updateTicket(ticket.id, projectId, { plan_ready: true }).catch(() => {})
+            }
+            get().moveTicket(ticket.id, projectId, 'review', ticket.sort_order).catch(() => {})
+          } else if (ticket.column === 'done' || ticket.column === 'merged') {
+            // Recover an explicit follow-up reopen this project missed while
+            // unloaded. Only while the session is still actively running — a
+            // session that already finished (or never started) gives no
+            // license to leave the terminal column.
+            if (!recoveredThisPass.has(ticket.current_session_id)) continue
+            if (status !== 'working' && status !== 'planning') continue
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts]: Fixed in 8c942e23 — reconcile recovery now also accepts answering / permission / command_approval, so a recovered run blocked on user input still reopens its terminal ticket instead of the handled-mark permanently consuming the pending reopen. Covered by a regression test.

File: src/renderer/src/stores/useWorktreeStatusStore.ts [hunk 15]
```diff
@@ -148,7 +174,29 @@ export const useWorktreeStatusStore = create<WorktreeStatusState>((set, get) =>
...
+        }
+        explicitSend =
+          oneShotMarker || hasUnconsumedSend || metadata?.hookEventName === 'UserPromptSubmit'
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useWorktreeStatusStore.ts]: Fixed in 8c942e23 — the explicit-send derivation now accepts reason 'claude_cli_plan_followup', so transcript-watcher-detected plan feedback reopens a done/merged ticket even when the UserPromptSubmit hook is delayed or unavailable. Covered by a regression test.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 16]
```diff
@@ -2219,6 +2227,35 @@ function PlanReviewModeContent({
...
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx:2259]: Fixed in 04889fdf — the attempt now also captures the session-status entry timestamp written by its own optimistic setSessionStatus. Every send, hook event, or auto-resume rewrites that entry, so implementAttemptSuperseded() detects any interleaved session transition (including event-driven moves that never touch messageSendTimes) and the late failure stands down entirely.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 17]
```diff
@@ -2219,6 +2227,35 @@ function PlanReviewModeContent({
...
+      useKanbanStore
+        .getState()
+        .moveTicket(ticket.id, ticket.project_id, columnBeforeImplement, ticket.sort_order)
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx]: Fixed in a5b2bb3a — moveTicket gained a skipCompletionEffects option that suppresses the terminal-move side effects (dependent auto-launch and the follow-up trigger), and the implement-failure rollback is its only caller, so restoring done/merged can no longer re-launch a dependent holding pending_launch_config.

File: src/renderer/src/components/kanban/KanbanTicketModal.tsx [hunk 18]
```diff
@@ -2219,6 +2227,43 @@ function PlanReviewModeContent({
...
+    // ticket back under it.
+    let sendStampAtImplement: number | undefined
+    let statusStampAtImplement: number | undefined
+    const implementAttemptSuperseded = (): boolean => {
+      const sessionId = ticket.current_session_id
+      if (sessionId == null) return false
+      if (messageSendTimes.get(sessionId) !== sendStampAtImplement) return true
+      const entry = useWorktreeStatusStore.getState().sessionStatuses[sessionId]
+      return entry?.timestamp !== statusStampAtImplement
+    }
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/components/kanban/KanbanTicketModal.tsx:2249]: Fixed in 8a6684e6 — the supersession check now captures and compares the session-status entry by object identity instead of timestamp. Every setSessionStatus creates a fresh entry object, so any newer transition is detected regardless of millisecond collisions.

File: src/renderer/src/stores/useKanbanStore.ts [hunk 19]
```diff
@@ -1154,6 +1244,21 @@ export const useKanbanStore = create<KanbanState>()(
             }
           }
         }
+        // Remember explicit sends so projects whose tickets weren't loaded
+        // yet can recover the reopen on load (reconcileFinishedSessions).
+        // The currently-loaded projects have already handled the send live.
+        if (event.type === 'session_working' && event.explicitSend === true) {
+          pendingExplicitReopens.set(sessionId, new Set(allTickets.keys()))
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts:1270]: Fixed in 1ec96fe8 — implement events now populate a parallel pending registry with the same lifecycle as explicit sends (loaded projects marked handled live, entries cleared on run end, attributed once per project by reconcileFinishedSessions), and the auto_approve_plan exclusion is applied at reconcile time when the loaded ticket can actually be checked. Covered by three regression tests (recovery, armed exclusion, run-end clearing).

File: src/renderer/src/stores/useKanbanStore.ts [hunk 20]
```diff
@@ -1154,6 +1262,26 @@ export const useKanbanStore = create<KanbanState>()(
...
         }
+        // Remember explicit sends and plan approvals so projects whose
+        // tickets weren't loaded yet can recover the reopen on load
+        // (reconcileFinishedSessions). The currently-loaded projects have
+        // already handled the event live.
+        if (event.type === 'session_working' && event.explicitSend === true) {
+          pendingExplicitReopens.set(sessionId, new Set(allTickets.keys()))
+        }
+        if (event.type === 'implement') {
+          pendingImplementReopens.set(sessionId, new Set(allTickets.keys()))
```
[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts:1273]: Fixed in 64490d53 — the implement failure paths now call clearPendingSessionReopens alongside the status clear and column rollback, dropping both pending registries for the session (the failed attempt registered them and no run-end event would ever clean them up). Covered by a regression test.

[the remark this answers is not in the document]
Comment (morapelker) [reply src/renderer/src/stores/useKanbanStore.ts:1273]: Fixed in 0abbe87c — went with the central version of your suggestion: clearSessionStatus / setSessionStatus(null) now emit a status_cleared coordination event that drops both pending registries, so every cancellation path (queued-dispatch failure in the drains, interrupts, teardown, failed implements) is covered by one rule — pending recoveries never outlive the status entry whose run created them. The per-site clearing added in 64490d53 was replaced by this. Transient clears inside implement/reject flows re-register via their own subsequent events, so ordering is preserved.
